echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/monitor/WebhookEventMonitor.java

```diff
@@ -0,0 +1,140 @@
+
+  private final PipelineCache pipelineCache;
+
+  @Autowired
+  public WebhookEventMonitor(@NonNull PipelineCache pipelineCache,
+                             @NonNull Action1<Pipeline> subscriber,
+                             @NonNull Registry registry) {
+    super(subscriber, registry);
+    this.pipelineCache = pipelineCache;
+  }
+
+  @Override
+  public void processEvent(Event event) {
+    super.validateEvent(event);
+    if (event.getDetails().getType() == null) {
+      return;
+    }
+
+    /* Need to create WebhookEvent, since TriggerEvent is abstract */
+    WebhookEvent webhookEvent = objectMapper.convertValue(event, WebhookEvent.class);
+    webhookEvent.setDetails(event.getDetails());
+    webhookEvent.setPayload(event.getContent());
+
+    Observable.just(webhookEvent)
+      .doOnNext(this::onEchoResponse)
+      .subscribe(triggerEachMatchFrom(pipelineCache.getPipelines()));
+  }
+
+  @Override
+  protected boolean isSuccessfulTriggerEvent(final TriggerEvent event) {
+    return true;
+  }
+
+  @Override
+  protected Function<Trigger, Pipeline> buildTrigger(Pipeline pipeline, TriggerEvent event) {
+    return trigger -> pipeline.withTrigger(trigger.atConstraints(event.getPayload()));
+  }
+
+  @Override
+  protected boolean isValidTrigger(final Trigger trigger) {
+    boolean valid =  trigger.isEnabled() &&
+      (
+          trigger.getType() != null
+      );
+
+    return valid;
+  }
+
+  @Override
+  protected Predicate<Trigger> matchTriggerFor(final TriggerEvent event) {
+    String type = event.getDetails().getType();
+
+    return trigger ->
+      trigger.getType().equals(type) &&
+        (
+          // The Constraints in the Trigger could be null. That's OK.
+          trigger.getConstraints() == null ||
+
+            // If the Constraints are present, check that there are equivalents in the webhook payload.
+            (  trigger.getConstraints() != null &&
+               isConstraintInPayload(trigger.getConstraints(), event.getPayload())
+            )
+
+        );
+
+  }
+
+  /**
+   * Check that there is a key in the payload for each constraint declared in a Trigger.
+   * Also check that if there is a value for a given key, that the value matches the value in the payload.
+   * @param constraints A map of constraints configured in the Trigger (eg, created in Deck).
+   * @param payload A map of the payload contents POST'd in the Webhook.
+   * @return Whether every key (and value if applicable) in the constraints map is represented in the payload.
+     */
+  protected boolean isConstraintInPayload(final Map constraints, final Map payload) {
+    for (Object key : constraints.keySet()){
+      if (!payload.containsKey(key) || payload.get(key) == null) {
+        log.info("Webhook trigger ignored. Item " + key.toString() + " was not found in payload");
+        return false;
+      }
+      if (!constraints.get(key).equals("") && (!constraints.get(key).equals(payload.get(key)))){
```

Comment:
If the payload is a hierarchical map, how should this work? Or are we thinking that the only valuable bits to constrain on will be at the root?